core/src/avm2/optimize.rs

```diff
@@ -0,0 +1,893 @@
+            Some(this_class)
+        } else {
+            None
+        }
+    } else {
+        None
+    };
+
+    // TODO: Store these argument types somewhere on the function so they don't
+    // have to be re-resolved every function call
+    let mut argument_types = Vec::new();
+    for argument in &method.signature {
+        let type_name = &argument.param_type_name;
+
+        let argument_type = if !type_name.has_lazy_component() {
+            activation
+                .domain()
+                .get_class(type_name, activation.context.gc_context)
+        } else {
+            None
+        };
+
+        argument_types.push(argument_type);
+    }
+
+    // Initial set of local types
+    let mut initial_local_types = Locals::new(method_body.num_locals as usize);
+    if let Some(this_class) = this_class {
+        initial_local_types.set_class_object(0, this_class);
+    }
+
+    let mut i = 1;
+    for argument_type in argument_types {
+        if let Some(argument_type) = argument_type {
+            initial_local_types.set_class(i, argument_type);
+        }
+        i += 1;
+    }
+
+    // Logic to only allow for type-based optimizations on types that
+    // we're absolutely sure about- invalidate the local register's
+    // known type if any other register-modifying opcodes mention them
+    // anywhere else in the function.
+    for op in &*code {
+        match op {
+            Op::SetLocal { index }
+            | Op::Kill { index }
+            | Op::IncLocal { index }
+            | Op::IncLocalI { index }
+            | Op::DecLocal { index }
+            | Op::DecLocalI { index } => {
+                if (*index as usize) < initial_local_types.len() {
+                    initial_local_types.set_any(*index as usize);
+                }
+            }
+            Op::HasNext2 {
+                object_register,
+                index_register,
+            } => {
+                if (*object_register as usize) < initial_local_types.len() {
+                    initial_local_types.set_any(*object_register as usize);
+                }
+                if (*index_register as usize) < initial_local_types.len() {
+                    initial_local_types.set_any(*index_register as usize);
+                }
+            }
+            _ => {}
+        }
+    }
+
+    let mut stack = Stack::new();
+
+    macro_rules! stack_pop_multiname {
+        ($index: expr) => {{
+            let multiname = method
+                .translation_unit()
+                // note: ideally this should be a VerifyError here or earlier
+                .pool_maybe_uninitialized_multiname(*$index, &mut activation.context);
+
+            if let Ok(multiname) = multiname {
+                stack.pop_for_multiname(multiname);
+                Some(multiname)
+            } else {
+                None
+            }
+        }};
+    }
+
+    let mut scope_stack = Stack::new();
```

Comment:
Doesn't appear to actually be used yet for anything yet - do we need this?

Reply:
No, but there's only three ops using it and this will come in useful for optimizing traits found by `findproperty` and `findpropstrict`